pkgs/top-level/all-packages.nix

```diff
@@ -22656,7 +22656,7 @@ in
 
   pianobooster = qt5.callPackage ../applications/audio/pianobooster { };
 
-  picard = callPackage ../applications/audio/picard { };
+  picard = python3Packages.callPackage ../applications/audio/picard { };
```

Comment:
please don't use this pattern, it does not compose. If Qt has a set of packages, and Python, which `callPackage` to use?

Reply:
So @FRidh you suggest to leave it with `callPackage`? I kind of like this because this is an elegant way to tell the expression to use whatever `qt5` is used by `python3Packages`.

Reply:
> please don't use this pattern, it does not compose. If Qt has a set of packages, and Python, which callPackage to use?

Nothing composes here. You have to be explicit in some cases. With my solution, you have to be explicit in less cases.

The question is not resolved by using `libsForQt5*.callPackage` or `callPackage ({ qt5, ... }: ... )`. Also, you'd have to find and change every package using python and qt again when python is updated to use qt515, which is not necessary if you use `python*Packages.callPackage`.

Somehow, you must prioritize which package set chooses the other package sets.
In general it is more likely that python is overriding qt than qt overriding python. So, the only quick and not completely dirty solution I see is exactly this pattern.

Reply:
> So @FRidh you suggest to leave it with callPackage? I kind of like this because this is an elegant way to tell the expression to use whatever qt5 is used by python3Packages.

Always use the `callPackage` from the set you are in.

Reply:
Why not define python applications in `pythonPackages` then?

Reply:
Using the `callPackage` of a sub-scope is officially recommended by the nixpkgs manual:

> **Adding an application to Nixpkgs.**  Add a Qt application to all-packages.nix using libsForQt5.callPackage instead of the usual callPackage. The former ensures that all dependencies are built with the same version of Qt.
– https://nixos.org/manual/nixpkgs/stable/#sec-language-qt

Reply:
> you'd have to find and change every package using python and qt again when python is updated to use qt515, which is not necessary if you use python*Packages.callPackage

Is this a pracitcal scenario? All the packages will not fail at build time but produce runtime errors!

Reply:
> Always use the `callPackage` from the set you are in.

@FRidh I'm afraid that's a hard requirement to comply to. To my view it contradicts a different requirements that you (`pythonPackages`' maintainers) also have - which is "don't put python applications in `pkgs/top-level/python-modules/` and `pkgs/top-level/python-packages.nix`".

Reply:
Right. Applications are in top-level, libraries in `python-packages.nix`. Libraries have Python packages as parameters, applications do not. The messy part is when a package is to be used as both, in which case library usage takes precedence, and one uses `foo = with python3Packages; toPythonApplication mypackage;`.